File: leetcode/1338-reduce-array-size-to-the-half/solution.cpp
int minSetSize(vector<int>& arr) {
    unordered_map<int, int> counts;

    for (int num : arr) {
        counts[num]++;
    }

    vector<int> ordered;

    for (auto [key, val] : counts) {
        ordered.push_back(val);
    }

    sort(ordered.begin(), ordered.end(), greater<int>());

    int current_size = arr.size();
    int removed_count = 0;
    for (int num : ordered) {
        if (current_size > (arr.size() / 2)) {
            current_size -= num;
            removed_count++;
        } else {
            break;
        }
    }

    return removed_count;
}

File: leetcode/1338-reduce-array-size-to-the-half/solution.ts
function minSetSize(arr: number[]): number {
  const counts = new Map<number, number>();

  for (const num of arr) {
    counts.set(num, (counts.get(num) || 0) + 1);
  }

  const ordered: number[] = [];

  for (const [key, value] of counts) {
    ordered.push(value);
  }

  ordered.sort(function (a: number, b: number): number {
    return b - a;
  });

  let currentSize = arr.length;
  let removedCount = 0;

  for (const num of ordered) {
    if (currentSize > Math.floor(arr.length / 2)) {
      currentSize -= num;
      removedCount++;
    } else {
      break;
    }
  }

  return removedCount;
}
